addons/hr_holidays/models/resource.py

```diff
@@ -80,8 +82,54 @@ def _reevaluate_leaves(self, time_domain_dict):
 
         leaves_to_cancel._force_cancel(_("a new public holiday completely overrides this leave."), 'mail.mt_comment')
 
+    def _convert_timezone(self, utc_naive_datetime_from, tz_from, tz_to):
+        """
+            Convert a naive date to another timezone that initial timezone
+            used to generate the date.
+            :param utc_naive_datetime: utc date without tzinfo
+            :type utc_naive_datetime: datetime
+            :param tz_from: timezone used to obtained `utc_naive_datetime`
+            :param tz_to: timezone in which we want the date
+            :return: datetime converted into tz_to without tzinfo
+            :rtype: datetime
+        """
+        naive_datetime_from = utc_naive_datetime_from.astimezone(tz_from).replace(tzinfo=None)
+        aware_datetime_to = tz_to.localize(naive_datetime_from)
+        utc_naive_datetime_to = aware_datetime_to.astimezone(pytz.utc).replace(tzinfo=None)
+        return utc_naive_datetime_to
+
+    def _ensure_datetime(self, datetime_representation, date_format=None):
```

Comment:
Yes, if you want your string containing a date to become a datetime, you need to pass a date format as a parameter to the function.

Yes, if you want your string containing a date to become a datetime, you need to pass a date format as a parameter to the function.

So if it is not a `datetime` or a `string` (plus the format) object, the function returns whatever is passed to it.

This behaviour prevents the function from becoming 'blocked' if it is used elsewhere. The only case I can think of that could happen is if you pass an object of type `date` and the function returns an object of type `date`. But in our flow, this will never be the case thanks to the condition which tests the types.

Therefore, the docstring and the name of the parameter clearly indicate that we are waiting for a representation of a `datetime` object.

We can possibly return None if we don't detect a datetime representation, what do you think?